include/zephyr/drivers/can/internal/can_impl.h

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2021 Vestas Wind Systems A/S
+ * Copyright (c) 2018 Karsten Koenig
+ * Copyright (c) 2018 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief Inline syscall implementation for Controller Area Network (CAN) driver APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CAN_H_
+#error "Should only be included by zephyr/drivers/can.h"
+#endif
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CAN_INTERNAL_CAN_IMPL_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CAN_INTERNAL_CAN_IMPL_H_
+
+#include <errno.h>
+
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <string.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/sys/util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static inline int z_impl_can_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_driver_api *api = (const struct can_driver_api *)dev->api;
```

Comment:
What make me a bit worried is that you are using the api struct here without including the definition, but only rely on the include order.
On the other hand, including the api definition header forms a recursion.
I'm not 100% sure is it is a good ida to split out the impl functions.
Is there a technical benefit from having them in a separate file?

Reply:
This is mainly to de-clutter the public header files so there is less "noise" there.

Also, these new internal header files should not be included directly, but only included by the main public header files. So not including the API definition there is kind of a subtle way to achieve, as it will generate compilation errors.